File: internal/templates/templates/todo-api/_node/services/complete-todo/handler.mjs
// POST /todos/{id}/done — mark one todo done (404 if it doesn't exist).
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, UpdateCommand } from "@aws-sdk/lib-dynamodb";

const db = DynamoDBDocumentClient.from(new DynamoDBClient({}));

export const handler = async (event) => {
  const id = event.pathParameters.id;
  try {
    const { Attributes } = await db.send(new UpdateCommand({
      TableName: process.env.TABLE_NAME,
      Key: { id },
      UpdateExpression: "SET done = :d",
      ExpressionAttributeValues: { ":d": true },
      ConditionExpression: "attribute_exists(id)", // 404 instead of upsert
      ReturnValues: "ALL_NEW",
    }));
    return { statusCode: 200, body: JSON.stringify(Attributes) };
  } catch (err) {
    if (err.name === "ConditionalCheckFailedException") {
      return { statusCode: 404, body: JSON.stringify({ error: `todo ${id} not found` }) };
    }
    throw err;
  }
};

File: internal/templates/templates/todo-api/_python/services/complete-todo/handler.py
"""POST /todos/{id}/done — mark one todo done (404 if it doesn't exist)."""

import json
import os

import boto3
from botocore.exceptions import ClientError

table = boto3.resource("dynamodb").Table(os.environ["TABLE_NAME"])


def handler(event, context):
    todo_id = event["pathParameters"]["id"]
    try:
        resp = table.update_item(
            Key={"id": todo_id},
            UpdateExpression="SET done = :d",
            ExpressionAttributeValues={":d": True},
            ConditionExpression="attribute_exists(id)",  # 404 instead of upsert
            ReturnValues="ALL_NEW",
        )
    except ClientError as e:
        if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
            return {"statusCode": 404, "body": json.dumps({"error": f"todo {todo_id} not found"})}
        raise

    return {"statusCode": 200, "body": json.dumps(resp["Attributes"], default=str)}
